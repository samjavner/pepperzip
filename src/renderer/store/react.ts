import * as React from "react";
import { Commands, Dispatch, Store, Update, UseStoreOptions } from "./";

export interface Msg<M extends object, U extends Update<M>> {
    type: keyof U;
    payload?: any;
}

export function useLocalStore(log: (msg: any) => void = () => ({})) {
    return function actual<
        M extends object,
        U extends Update<M>,
        C extends Commands = {}
    >({
        init,
        update,
        createCommands,
        memo = [],
    }: UseStoreOptions<M, U, C>): Store<M, U, C> {
        const [model, reducerDispatch] = React.useReducer(
            (state: M, action: Msg<M, U>) =>
                update[action.type](state, action.payload),
            init
        );

        const vm = React.useMemo(
            () => {
                const dispatch: Dispatch<U> = {} as any;
                Object.keys(update).forEach(type => {
                    (dispatch as any)[type] = (payload: any) => {
                        const msg =
                            update[type].length === 1
                                ? { type }
                                : { type, payload };
                        log(msg);
                        reducerDispatch(msg);
                    };
                });
                const commands: C = createCommands
                    ? createCommands(model, dispatch)
                    : (({} as any) as C);
                return {
                    ...model,
                    ...dispatch,
                    ...commands,
                };
            },
            [model, update, ...memo]
        );

        return vm;
    };
}
