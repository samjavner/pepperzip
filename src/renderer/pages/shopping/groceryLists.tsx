import * as React from "react";

// VIEW

export const Page: React.FunctionComponent = () => <div>Grocery Lists</div>;
