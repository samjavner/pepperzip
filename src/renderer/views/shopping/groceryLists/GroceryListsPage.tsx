import * as React from "react";
import { connect } from "react-redux";
import { Dispatch } from "redux";
import * as actions from "../../../app/actions";
import { State } from "../../../app/state";

function mapStateToProps(state: State) {
    return {};
}

function mapDispatchToProps(dispatch: Dispatch<actions.Action>) {
    return {};
}

type GroceryListsPageProps = ReturnType<typeof mapStateToProps> &
    ReturnType<typeof mapDispatchToProps>;

export const GroceryListsPage: React.SFC<GroceryListsPageProps> = props => (
    <div>Grocery Lists</div>
);

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(GroceryListsPage);
