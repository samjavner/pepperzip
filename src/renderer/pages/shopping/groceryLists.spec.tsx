import * as React from "react";
import * as renderer from "react-test-renderer";
import * as groceryLists from "./groceryLists";

describe("groceryLists", () => {
    describe("Page", () => {
        it("renders the text 'GroceryLists'", () => {
            const component = renderer.create(<groceryLists.Page />);
            const tree = component.toJSON();
            expect(tree).toMatchSnapshot();
        });
    });
});
