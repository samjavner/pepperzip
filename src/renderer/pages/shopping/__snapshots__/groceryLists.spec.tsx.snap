// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`groceryLists Page renders the text 'GroceryLists' 1`] = `
<div>
  Grocery Lists
</div>
`;
